### ISSUE TITLE
Unnecessary cast diagnostic sometimes missing in top level programs

### ISSUE THREAD

**Issue by dibarbet** [dotnet/roslyn#60248]
**Version Used**:  17.1.1 (also repros in main)

**Steps to Reproduce**:

1.  Create a new .net 6 console app
2.  Replace the contents of program.cs with 
```
int x = 1;
int y = (int)x;
int z = (int)1;

class C
{
    void M()
    {
        int i = 1;
        int j = (int)j;
        int k = (int)1;
    }
}
```
3.  Note the unnecessary cast diagnostic only on the cast in `z` but not on `y`.  Note that both unnecessary casts in method `M()` are marked as unnecessary (`j` and `k`).

**Expected Behavior**:
An unnecessary cast diagnostic on `y` as well.

**Comment by Youssef1313** [comment]
I don't see a bug here. The missing diagnostic on `int y = (int)j;` sounds correct since `j` is not declared and we don't know its type.

**Comment by dibarbet** [comment]
> I don't see a bug here. The missing diagnostic on `int y = (int)j;` sounds correct since `j` is not declared and we don't know its type.

Good catch, was a typo in the repro, it should say x
.e.g
![image](https://user-images.githubusercontent.com/5749229/159055390-b8f8f863-d8a0-42f1-9184-72a0d0738470.png)
